app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php

```diff
@@ -0,0 +1,861 @@
+use bandwidthThrottle\tokenBucket\storage\StorageException;
+use bandwidthThrottle\tokenBucket\TokenBucket;
+use Exception;
+use Joomla\Http\Exception\UnexpectedResponseException;
+use Joomla\Http\Http;
+use Mautic\EmailBundle\Model\TransportCallback;
+use Mautic\EmailBundle\MonitoredEmail\Exception\BounceNotFound;
+use Mautic\EmailBundle\MonitoredEmail\Exception\UnsubscriptionNotFound;
+use Mautic\EmailBundle\MonitoredEmail\Message;
+use Mautic\EmailBundle\MonitoredEmail\Processor\Bounce\BouncedEmail;
+use Mautic\EmailBundle\MonitoredEmail\Processor\Bounce\Definition\Category;
+use Mautic\EmailBundle\MonitoredEmail\Processor\Bounce\Definition\Type;
+use Mautic\EmailBundle\MonitoredEmail\Processor\Unsubscription\UnsubscribedEmail;
+use Mautic\LeadBundle\Entity\DoNotContact;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\HttpException;
+use Symfony\Component\Translation\TranslatorInterface;
+
+/**
+ * Class AmazonApiTransport.
+ */
+class AmazonApiTransport extends AbstractTokenArrayTransport implements \Swift_Transport, TokenTransportInterface, CallbackTransportInterface, BounceProcessorInterface, UnsubscriptionProcessorInterface
+{
+    /**
+     * From address for SNS email.
+     */
+    const SNS_ADDRESS = 'no-reply@sns.amazonaws.com';
+
+    /**
+     * @var string
+     */
+    private $region;
+
+    /**
+     * @var string
+     */
+    private $username;
+
+    /**
+     * @var string
+     */
+    private $password;
+
+    /**
+     * @var int
+     */
+    private $concurrency;
+
+    /**
+     * @var SesClient
+     */
+    private $client;
+
+    /**
+     * @var Http
+     */
+    private $httpClient;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var TransportCallback
+     */
+    private $transportCallback;
+
+    /**
+     * @var BlockingConsumer
+     */
+    private $createTemplateBucketConsumer;
+
+    /**
+     * @var BlockingConsumer
+     */
+    private $sendTemplateBucketConsumer;
+
+    /**
+     * @var array
+     */
+    private $templateCache;
+
+    /**
+     * AmazonApiTransport constructor.
+     */
+    public function __construct(Http $httpClient, LoggerInterface $logger, TranslatorInterface $translator, TransportCallback $transportCallback)
+    {
+        $this->logger            = $logger;
+        $this->translator        = $translator;
+        $this->httpClient        = $httpClient;
+        $this->transportCallback = $transportCallback;
+
+        $this->templateCache     = [];
+    }
+
+    public function __destruct()
+    {
+        if (count($this->templateCache)) {
+            $this->logger->debug('Deleting SES templates that were created in this session');
+            foreach ($this->templateCache as $templateName) {
+                $this->deleteSesTemplate($templateName);
```

Comment:
https://github.com/mautic/mautic/pull/6977#issuecomment-519614524

Didn't tested this PR, but running #6977 this delete crashed the integration. I think AWS takes some time to process the emails and we remove the template before they finish processing.

I'm running without deleting templates. It is not perfect, but better than unexpected crashes.

```
[2019-08-08 17:02:55] mautic.ERROR: Exception sending email template: TemplateDoesNotExist, Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist [] []
[2019-08-08 17:02:55] mautic.ERROR: [MAIL ERROR] Error executing "SendBulkTemplatedEmail" on "https://email.us-east-1.amazonaws.com"; AWS HTTP error: Client error: `POST https://email.us-east-1.amazonaws.com` resulted in a `400 Bad Request` response: <ErrorResponse xmlns="http://ses.amazonaws.com/doc/2010-12-01/">   <Error>     <Type>Sender</Type>     <Code>TemplateDoe (truncated...)  TemplateDoesNotExist (client): Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist - <ErrorResponse xmlns="http://ses.amazonaws.com/doc/2010-12-01/">   <Error>     <Type>Sender</Type>     <Code>TemplateDoesNotExist</Code>     <Message>Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist</Message>   </Error>   <RequestId>fa6e5393-3d5b-4089-a3f3-7774213227a8</RequestId> </ErrorResponse>  (send); REDACTED_EMAILS_WERE_HERE {"exception":"[object] (Exception(code: 0): Error executing \"SendBulkTemplatedEmail\" on \"https://email.us-east-1.amazonaws.com\"; AWS HTTP error: Client error: `POST https://email.us-east-1.amazonaws.com` resulted in a `400 Bad Request` response:\n<ErrorResponse xmlns=\"http://ses.amazonaws.com/doc/2010-12-01/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>TemplateDoe (truncated...)\n TemplateDoesNotExist (client): Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist - <ErrorResponse xmlns=\"http://ses.amazonaws.com/doc/2010-12-01/\">\n  <Error>\n    <Type>Sender</Type>\n    <Code>TemplateDoesNotExist</Code>\n    <Message>Template MauticTemplate-93-810079ab445339df9a538d5519fc5383 does not exist</Message>\n  </Error>\n  <RequestId>fa6e5393-3d5b-4089-a3f3-7774213227a8</RequestId>\n</ErrorResponse>\n at /var/www/html/app/bundles/EmailBundle/Swiftmailer/Transport/AmazonApiTransport.php:457)"} []
``

Reply:
With #6977 we did not have to remove these lines. Works fine for me. We had to add a pid to fix the error mentioned. From what the Author and I gathered, it seemed 2 processes were sending emails in parallel, and a race condition occurred which deleted the template before the second process sent emails, but after the second process found the templates to already exist.

Reply:
Great! Glad you already investigated and handled the issue. I'll try this solution in my production code.

Reply:
@heroic may you please add the lines that solved this issue?

Reply:
Guess you are refering to adding the getmypid() to the template name like what i did in line 502, right?

Reply:
@Wise-Gorilla yes.

Reply:
https://github.com/mautic/mautic/pull/9005/files#diff-327931ffb6a20b14b8da54a78284787cR502

Reply:
@Wise-Gorilla @mabumusa1 was this resolved? Please resolve the conversation if so.

Reply:
Looks like this was resolved. I'm resolving the conversation so we can go ahead and merge 👍